dora/services/tests.py

```diff
@@ -1914,3 +1930,398 @@ def test_get_previous_status_returns_correct_value(self):
+        self.assertEqual(
+            ServiceCategory.objects.filter(value__in=[value, value_2]).count(), 2
+        )
+
+        # QUAND je supprime une thématique
+        delete_category(ServiceCategory, value)
+
+        # ALORS il en reste toujours une
+        self.assertEqual(
+            ServiceCategory.objects.filter(value__in=[value, value_2]).count(), 1
+        )
+        self.assertEqual(ServiceCategory.objects.filter(value=value_2).count(), 1)
+
+    def test_delete_subcategory(self):
+        # ÉTANT DONNÉ un besoin
+        value = "test-subcategory"
+        baker.make("ServiceSubCategory", value=value, label="Label_1")
+        self.assertEqual(ServiceSubCategory.objects.filter(value=value).count(), 1)
+
+        # QUAND je supprime ce besoin
+        delete_subcategory(ServiceSubCategory, value)
+
+        # ALORS il n'existe plus
+        self.assertEqual(ServiceSubCategory.objects.filter(value=value).count(), 0)
+
+    def test_delete_two_subcategories(self):
+        # ÉTANT DONNÉ deux besoins
+        value = "test-subcategory"
+        baker.make("ServiceSubCategory", value=value, label="Label_1")
+
+        value_2 = "test-subcategory-2"
+        baker.make("ServiceSubCategory", value=value_2, label="Label_2")
+        self.assertEqual(
+            ServiceSubCategory.objects.filter(value__in=[value, value_2]).count(), 2
+        )
+
+        # QUAND je supprime ce besoin
+        delete_subcategory(ServiceSubCategory, value)
+
+        # ALORS il en reste toujours un
+        self.assertEqual(
+            ServiceSubCategory.objects.filter(value__in=[value, value_2]).count(), 1
+        )
+        self.assertEqual(ServiceSubCategory.objects.filter(value=value_2).count(), 1)
+
+    def test_replace_subcategory_err_to_category_not_exists(self):
+        # ÉTANT DONNÉ une thématique existante et une thématique indexistante
+        value = "subcategory_1"
+        subcategory = baker.make("ServiceSubCategory", value=value)
+        service = make_service()
+        service.subcategories.add(subcategory)
+
+        # QUAND je remplace la thématique existante par une non existante
+        try:
+            replace_subcategory(
+                ServiceSubCategory, Service, value, "non_existing_subcategory"
+            )
+        except Exception as e:
+            err = e
+
+        # ALORS j'obtiens une erreur
+        self.assertTrue(isinstance(err, ValidationError))
+
+    def test_replace_subcategory(self):
+        # ÉTANT DONNÉ deux besoins
+        subcategory_value_1 = "subcategory_1"
+        subcategory_1 = baker.make("ServiceSubCategory", value=subcategory_value_1)
+        service = make_service()
+        service.subcategories.add(subcategory_1)
+
+        subcategory_value_2 = "subcategory_2"
+        baker.make("ServiceSubCategory", value=subcategory_value_2)
+
+        service = Service.objects.filter(pk=service.pk).first()
+        subcategories = [s["value"] for s in service.subcategories.values()]
```

Comment:
ok tu as souvent utilisé ça dans la PR principale aussi, mais ça a peu de sens (même si ça m'arrive de le faire aussi…).
- Si tu veux récuperer une liste, le plus simple est .values_list avec flat=True
- Si pour une raison étrange tu n'aimes pas values_list, ok pour un compréhension, mais aucune raison de passer par un dictionnaire intermédiaire:
`[s.value for s in service.subcategories.all()]`

Reply:
c'était surtout l'argument `flat` qui me gênait
mais je me suis renseigné un peu ( https://stackoverflow.com/a/37205928/9308644) et c'est un peu plus clair ^^